File: projects/17-global-conferencing/src/cascade.rs
//! V2 — Inter-SFU cascade transport. `src/cascade.rs`.
//!
//! This is the heart of "cascaded": the transport that makes an SFU a **peer of another SFU**. A
//! publisher's origin media lives in its home region (V1). A subscriber in a *different* region
//! must receive it — but the origin SFU must **not** send one copy per remote subscriber across
//! the ocean. Instead it opens **one relay leg per remote region with interest** and sends a
//! **single** copy of each forwarded stream down that leg; the remote SFU receives it and does the
//! **local** fan-out (reusing project 15's per-subscriber [`Rewriter`] for its own locals). Inter-
//! region cost becomes `O(regions)`, not `O(subscribers)` — the whole win.
//!
//! The subtleties are all about being a relay **without a loop**: a relay copy arriving from region
//! `A` is fanned out to local subscribers but is **never** relayed onward (to `A` or a third
//! region), or a 3-region mesh forwards a packet in circles forever. So a relayed packet carries
//! provenance, the SFU relays **only origin media** and fans out **relay media locally**, and legs
//! are **bounded** (a fixed peer set, one leg each, torn down when the last remote subscriber
//! leaves).
//!
//! Scaffold state: [`CascadeMesh::new`] and the read-only [`links`](CascadeMesh::links) snapshot are
//! wired so `/status` shows the (empty) leg table. Opening a leg, relaying a copy out, and fanning a
//! relayed copy in are the V2 `todo!()` worklist. The backbone UDP socket is bound in `main` and
//! handed to [`run`](CascadeMesh::run).

use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};

use bytes::Bytes;
use serde::Serialize;
use tokio::net::UdpSocket;

use crate::error::Result;
use crate::placement::PeerNode;

/// One backbone relay leg: a live SFU→SFU link carrying origin media to a peer region that has
/// interest. Exists only while that region has ≥1 subscriber for a stream here.
#[derive(Clone, Debug, Serialize)]
pub struct RelayLink {
    /// The peer region this leg feeds.
    pub region: String,
    /// Where relay copies are sent (the peer's cascade UDP address).
    pub remote_addr: String,
    /// How many distinct origin streams (publisher tracks) are currently relayed on this leg.
    pub tracks: usize,
}

/// A relayed datagram to send on the backbone socket (analogous to project 15's `Outgoing`).
#[derive(Debug)]
pub struct RelayOut {
    pub dst: SocketAddr,
    pub data: Bytes,
}

/// Config for the cascade mesh, read from env in `main`.
pub struct CascadeConfig {
    /// This SFU's region — the loop guard: a relay copy is never sent back to, or fanned out as if
    /// from, this region, and origin media from here is relayed out but never re-relayed.
    pub region: String,
    /// The backbone UDP port this SFU relays on.
    pub cascade_port: u16,
    /// The peer SFUs (region → cascade address) legs can be opened to.
    pub peers: Vec<PeerNode>,
    /// Cap on concurrently-open relay legs (abuse/backpressure bound).
    pub max_links: usize,
}

/// Mutable leg table, guarded by one lock (sync work only; the socket `send_to` awaits happen in
/// `run`/the caller after the lock is dropped — the p15 locking discipline).
#[derive(Default)]
struct Inner {
    /// Open relay legs keyed by peer region.
    links: HashMap<String, RelayLink>,
}

/// The cascade mesh: the set of relay legs to peer regions + the backbone transport.
pub struct CascadeMesh {
    cfg: CascadeConfig,
    inner: Mutex<Inner>,
}

impl CascadeMesh {
    /// Build the mesh. Wiring only — legs open on demand (V2), not here.
    pub fn new(cfg: CascadeConfig) -> Self {
        Self {
            cfg,
            inner: Mutex::new(Inner::default()),
        }
    }

    pub fn config(&self) -> &CascadeConfig {
        &self.cfg
    }

    /// This SFU's region (the loop guard).
    pub fn region(&self) -> &str {
        &self.cfg.region
    }

    /// A snapshot of the open relay legs (for `/status`).
    pub fn links(&self) -> Vec<RelayLink> {
        self.inner
            .lock()
            .expect("cascade lock")
            .links
            .values()
            .cloned()
            .collect()
    }

    /// The peer node for a region, if it's in the mesh (used to authenticate a relay source and to
    /// resolve where a leg sends). Wired lookup — no I/O.
    pub fn peer(&self, region: &str) -> Option<&PeerNode> {
        self.cfg.peers.iter().find(|p| p.region == region)
    }

    // ---- V2 worklist: open legs · relay out · fan in (loop-free) --------------------------

    /// TODO(V2): Ensure a relay leg to `region` exists for `stream` (idempotent). Opens the leg on
    /// the first remote subscriber for a stream there, bumps its track count, enforces
    /// `max_links`. Returns the leg. Called from signaling when a subscriber's publisher is in
    /// another region.
    pub async fn ensure_link(&self, region: &str, stream: &str) -> Result<RelayLink> {
        let _ = (region, stream, self.cfg.max_links);
        todo!("V2: open-or-reuse the relay leg to this region, enforce MAX_RELAY_LINKS")
    }

    /// TODO(V2): Tear down interest in `stream` on the leg to `region`; close the leg when its last
    /// track's last remote subscriber leaves. Idempotent (safe on an already-closed leg).
    pub async fn release_link(&self, region: &str, stream: &str) -> Result<()> {
        let _ = (region, stream);
        todo!("V2: decrement leg interest, close the leg when it reaches zero")
    }

    /// TODO(V2): Relay **one** copy of an origin packet for `stream` to each remote region with
    /// interest — one `send_to` per leg, regardless of how many subscribers that region has (it
    /// fans out locally). **Loop guard:** only *origin* media (produced in this region) is relayed;
    /// never re-relay a packet that arrived as a relay copy. Bumps `RELAY_COPIES_OUT` per region.
    pub async fn relay_out(&self, stream: &str, packet: &[u8]) -> Result<Vec<RelayOut>> {
        let _ = (stream, packet, self.cfg.region.as_str());
        todo!("V2: send one relay copy per interested remote region (origin media only)")
    }

    /// TODO(V2): A relay copy arrived from a peer SFU (`from`). Authenticate the source (must be a
    /// known peer region), then hand it to the **local** fan-out (project 15's rewriter, per local
    /// subscriber). **Loop guard:** a relayed copy is fanned out locally and **never** re-relayed
    /// onward. Bumps `RELAY_COPIES_IN`; drops unknown/looping/truncated packets by reason.
    pub async fn on_relayed(&self, from: SocketAddr, packet: &[u8]) -> Result<()> {
        let _ = (from, packet);
        todo!("V2: authenticate peer, fan out locally, never re-relay (loop-free)")
    }

    /// TODO(V2): The backbone pump — receive relay copies on the cascade socket and dispatch them
    /// to [`on_relayed`](Self::on_relayed), draining outbound relay copies via `socket.send_to`.
    /// Runs until `shutdown`. Gated behind `RUN_BACKGROUND` in `main` (a bare scaffold with no
    /// peers has nothing to relay).
    pub async fn run(
        self: Arc<Self>,
        socket: Arc<UdpSocket>,
        mut shutdown: tokio::sync::watch::Receiver<bool>,
    ) -> Result<()> {
        let _ = &socket;
        let _ = shutdown.changed().await;
        todo!("V2: recv relay copies -> on_relayed; send outbound relay copies; drain on shutdown")
    }
}

File: projects/17-global-conferencing/src/global_conferencing/cascade.py
"""V2 — Inter-SFU cascade transport. `src/global_conferencing/cascade.py`.

This is the heart of "cascaded": the transport that makes an SFU a **peer of
another SFU**. A publisher's origin media lives in its home region (V1). A
subscriber in a *different* region must receive it — but the origin SFU must
**not** send one copy per remote subscriber across the ocean. It opens **one
relay leg per remote region with interest** and sends a **single** copy of each
forwarded stream down that leg; the remote SFU receives it and does the **local**
fan-out through project 15's per-subscriber rewriter. Inter-region cost becomes
`O(regions × layers)`, not `O(subscribers)` — the whole win.

The subtleties are all about being a relay **without a loop**. A relay copy
arriving from region `A` is fanned out to local subscribers but **never**
relayed onward — not back to `A`, not to a third region — or a 3-region mesh
forwards one packet in circles forever, amplifying itself into a backbone storm.
So a relayed datagram carries **provenance**, the SFU relays **only origin
media** and fans **relay media** out locally, and legs are **bounded**: a fixed
peer set, one leg each, torn down when the last remote subscriber leaves.

## The framing is yours, and it is attack surface

What a relay datagram looks like on the backbone — which bytes say which track,
which layer, which region it originated in — is V2's design, and it goes in
`docs/17-design.md`. Whatever you pick, `on_relayed` parses bytes that arrived
on an **open UDP port**. `struct.unpack_from` raises `struct.error` on a short
buffer, but plain slicing does not raise at all, so check `len(datagram)`
against the header size *first* and raise `TruncatedError`; the backbone pump
catches `MediaError` and moves on. `struct.pack` into a fixed header plus the
RTP bytes is the direct way to build one; for `O(regions)` copies the
allocation is not your bottleneck — the per-*subscriber* copies on the far side
are.

Scaffold state: construction, the peer lookup and the `/status` leg table are
wired, and the backbone UDP socket is bound and pumped in `transport.py`. The
first datagram that reaches the backbone port calls `on_relayed` and raises; the
pump task ends and `/readyz` turns 503 while `/healthz` stays green.
"""

from __future__ import annotations

from dataclasses import dataclass

from pydantic import BaseModel

from .config import PeerNode
from .ids import Address, LayerId, TrackKey
from .routing import LayerRouter

__all__ = ["CascadeConfig", "CascadeMesh", "LocalDelivery", "RelayDatagram", "RelayLeg"]


class RelayLeg(BaseModel):
    """One backbone relay leg — a live SFU↔SFU link for a peer region with interest.

    Exists only while that region has ≥1 subscriber for a stream relayed on it.
    The `/rooms` JSON the web playground renders is a list of these.
    """

    region: str
    """The peer region this leg is with."""

    remote_addr: str
    """The peer's backbone `host:port`."""

    tracks: int
    """Distinct origin tracks currently relayed on this leg."""


@dataclass(frozen=True, slots=True)
class RelayDatagram:
    """One datagram to send on the backbone socket."""

    dst: Address
    payload: bytes


@dataclass(frozen=True, slots=True)
class LocalDelivery:
    """What a valid relay copy becomes on the far side: the input to local fan-out.

    `packet` is the RTP datagram as the origin forwarded it, with the relay
    framing removed — what project 15's rewriter consumes for each local
    subscriber of `track`.
    """

    track: TrackKey
    packet: bytes


@dataclass(frozen=True, slots=True)
class CascadeConfig:
    region: str
    """This SFU's region — the loop guard's notion of "origin media is from here"."""

    peers: tuple[PeerNode, ...]
    """The peer SFUs legs can exist with."""

    max_legs: int
    """Cap on concurrently open relay legs (`MAX_RELAY_LINKS`)."""


class CascadeMesh:
    """The set of relay legs to peer regions, and the relay/fan-in decisions on them."""

    def __init__(self, config: CascadeConfig, router: LayerRouter) -> None:
        self.config = config
        self._router = router
        """V3, consulted per packet: a leg carries only the layers its region demands."""

        self._peers = {peer.region: peer for peer in config.peers}
        self._legs: dict[str, RelayLeg] = {}
        """Open legs keyed by peer region."""

        # TODO(V2): the rest of the leg state. Roughly: per leg, *which* tracks it
        # relays (a `set[TrackKey]` — `RelayLeg.tracks` is its length, not the set),
        # and per track, which regions want it, so `relay_out` is one dictionary
        # lookup per packet rather than a scan of every leg. And a way to recognise
        # a known peer by the *address* a datagram came from.

    @property
    def region(self) -> str:
        return self.config.region

    def legs(self) -> list[RelayLeg]:
        """A snapshot of the open relay legs (for `/rooms` and `/status`)."""
        return [self._legs[region] for region in sorted(self._legs)]

    def peer(self, region: str) -> PeerNode | None:
        """The peer SFU for `region`, if it is in the mesh."""
        return self._peers.get(region)

    # ---- V2 worklist: open legs · relay out · fan in (loop-free) -----------

    async def ensure_leg(self, region: str, track: TrackKey) -> RelayLeg:
        """Ensure a relay leg with `region` exists for `track`, and return it.

        TODO(V2): idempotent open-or-reuse. The first remote interest in a track
        opens the leg (bump `RELAY_LEGS{peer=region}`); later interest in the same
        track changes nothing on the backbone — that is the "doubling remote
        subscribers adds zero backbone packets" criterion, enforced here. Raise
        `NotFoundError` for a region not in `PEERS` and `ConflictError` past
        `config.max_legs`.

        Async because the origin has to *learn* that this region wants the track —
        the subscriber is here, the media is there. Whether that intent travels
        over HTTP, on the backbone itself, or is derived from V1's replicated
        membership is your call. Called from signaling when a subscriber's
        publisher lives in another region.
        """
        raise NotImplementedError(
            "V2: open-or-reuse the relay leg to this region, enforce MAX_RELAY_LINKS"
        )

    async def release_leg(self, region: str, track: TrackKey) -> None:
        """Drop interest in `track` on the leg with `region`.

        TODO(V2): close the leg when its last track goes. Idempotent — releasing an
        already-closed leg is harmless, because a subscriber's leave and a region's
        membership retirement can both arrive for the same track.
        """
        raise NotImplementedError("V2: decrement leg interest, close the leg when it reaches zero")

    def relay_out(self, track: TrackKey, layer: LayerId, packet: bytes) -> list[RelayDatagram]:
        """Relay **one** copy of an origin packet to each remote region that wants it.

        TODO(V2): one `RelayDatagram` per leg carrying `track` *and* whose V3 set
        carries `layer` (`self._router.leg_carries`) — regardless of how many
        subscribers that region has. Frame it with provenance so the far side can
        tell it is a relay copy. **Loop guard:** this is only ever called with
        media produced in this region; a packet that arrived as a relay copy must
        never come back through here. Bump `RELAY_COPIES_OUT_TOTAL` and
        `RELAY_BYTES_OUT_TOTAL` per region pair.

        Synchronous on purpose: it runs per origin packet, and the returned
        datagrams are sent with a non-blocking `sendto`.
        """
        raise NotImplementedError(
            "V2: send one relay copy per interested remote region (origin media only)"
        )

    def on_relayed(self, source: Address, datagram: bytes) -> LocalDelivery | None:
        """A datagram arrived on the backbone socket from `source`.

        TODO(V2): authenticate the source (a known peer's backbone address, else
        raise `UnknownPeerError`), bounds-check and strip the framing (else
        `TruncatedError`), and return the `LocalDelivery` for local fan-out.
        **Loop guard:** a relay copy is delivered locally and **never** re-relayed
        — there is no code path from here to `relay_out`. Return `None` to drop a
        copy that would close a loop (bump `RELAY_DROPPED_TOTAL{reason="loop"}`).
        Bump `RELAY_COPIES_IN_TOTAL` for each accepted copy.
        """
        raise NotImplementedError(
            "V2: authenticate peer, strip framing, fan out locally, never re-relay (loop-free)"
        )

    async def close_all(self) -> None:
        """Tear down every relay leg on graceful shutdown.

        TODO(V2): no half-open legs — a peer still relaying to a stopped SFU sends
        media at a port nobody drains. Tell each peer (however `ensure_leg` told
        it), then clear the table. Called from the lifespan before the backbone
        socket closes.
        """
        raise NotImplementedError("V2: tear down every relay leg cleanly (graceful shutdown)")
